src/Reference.js

```diff
@@ -15,34 +16,38 @@ export default class Reference {
   }
 
   constructor(key, mapFn, options = {}) {
-    validateName(key);
+    validateType(key);
     let prefix = options.contextPrefix || '$';
 
-    if (typeof key === 'function') {
-      key = '.';
-    }
-
     this.key = key.trim();
     this.prefix = prefix;
     this.isContext = this.key.indexOf(prefix) === 0;
-    this.isSelf = this.key === '.';
-
+    this.isRelativePath = isRelativePath(this.key);
     this.path = this.isContext ? this.key.slice(this.prefix.length) : this.key;
-    this._get = getter(this.path, true);
+    if (!this.isRelativePath) {
+      this._get = getter(this.path, true);
+    }
     this.map = mapFn || (value => value);
   }
   resolve() {
     return this;
   }
 
-  cast(value, { parent, context }) {
-    return this.getValue(parent, context);
+  cast(value, options) {
+    return this.getValue(options);
   }
 
-  getValue(parent, context) {
-    let isContext = this.isContext;
-    let value = this._get(isContext ? context : parent || context || {});
-    return this.map(value);
+  getValue(options) {
+    const { context, parent, path, value = {} } = options;
+    let refValue;
+    if (this.isRelativePath) {
+      this.path = getRelativePath(path, this.key);
```

Comment:
I don't love attaching context specific properties to the instance, since schema are supposed to immutable. Can you talk through why you went with this approach?